File: D&C - Maximum Subarray Sum/Cpp/maxSubarraySum.cpp
/* 	Author:		Vishal Gaur
* 	Created:	06-26-2021  17:14:44
*/

#include <bits/stdc++.h>

using namespace std;

// function to find max crossing subarray
vector<int> findMaxCrossingSubarray(vector<int> A, int low, int high, int mid)
{
    int sum = 0;
    int leftSum = INT_MIN;
    int maxLeft = mid;

    for (int i = mid; i >= low; i--)
    {
        sum += A[i];
        if (sum > leftSum)
        {
            leftSum = sum;
            maxLeft = i;
        }
    }

    sum = 0;
    int rightSum = INT_MIN;
    int maxRight = mid;

    for (int i = mid+1; i <= high; i++)
    {
        sum += A[i];
        if (sum > rightSum)
        {
            rightSum = sum;
            maxRight = i;
        }
    }

    return {maxLeft, maxRight, leftSum + rightSum};
}

// function to find maximum subarray
vector<int> findMaxSubarray(vector<int> A, int low, int high)
{
    if (low == high)
    {
        return {low, high, A[low]};
    }
    else
    {
        vector<int> ans;

        int mid = (low + high) / 2;
        vector<int> leftAns = findMaxSubarray(A, low, mid);
        vector<int> rightAns = findMaxSubarray(A, mid + 1, high);
        vector<int> crossAns = findMaxCrossingSubarray(A, low, high, mid);

        if (leftAns[2] >= rightAns[2] && leftAns[2] >= crossAns[2])
            ans = leftAns;
        else if (rightAns[2] >= leftAns[2] && rightAns[2] >= crossAns[2])
            ans = rightAns;
        else
            ans = crossAns;

        return ans;
    }
}

// main function to test above function

int main()
{
    vector<int> arr = {1, -4, -8, 6, 9, -4, -5, 10};
    int len = arr.size();

    vector<int> res = findMaxSubarray(arr, 0, len - 1);

    cout << "Maximum Subarray Sum is " << res[2] << endl;
    cout << "And subarray is: ";
    for (int i = res[0]; i <= res[1]; i++)
        cout << arr[i] << "  ";
    cout << endl;

    return 0;
}

File: D&C - Maximum Subarray Sum/Python3/maxSubarraySum.py
# Author:   Vishal Gaur
# Created:  06-26-2021  17:45:12

import math

# function to find max crossing subarray


def findMaxCrossingSubarray(A, low, mid, high):
    sum = 0
    leftSum = -math.inf
    maxLeft = mid

    for i in range(mid, low, -1):
        sum += A[i]
        if sum > leftSum:
            leftSum = sum
            maxLeft = i

    sum = 0
    rightSum = -math.inf
    maxRight = mid+1
    for i in range(mid+1, high+1):
        sum += A[i]
        if sum > rightSum:
            rightSum = sum
            maxRight = i

    return maxLeft, maxRight, leftSum+rightSum

# function to find maximum subarray


def findMaxSubarray(A, low, high):
    if low == high:
        return low, high, A[low]

    mid = (low+high) // 2

    leftLow, leftHigh, leftSum = findMaxSubarray(A, low, mid)
    rightLow, rightHigh, rightSum = findMaxSubarray(A, mid+1, high)
    crossLow, crossHigh, crossSum = findMaxCrossingSubarray(A, low, mid, high)

    if leftSum >= rightSum and leftSum >= crossSum:
        return leftLow, leftHigh, leftSum
    elif rightSum >= leftSum and rightSum >= crossSum:
        return rightLow, rightHigh, rightSum
    else:
        return crossLow, crossHigh, crossSum


# Driver Code to test above function

arr = [103, -74, -38, 96, 19, -34, -55, 50, 123]
size = len(arr)

start, end, sum = findMaxSubarray(arr, 0, size-1)

print("Maximum Subarray Sum is ", sum)
print("Subrray is: ", arr[start:end+1])
